xrpl4j-client/src/main/java/org/xrpl/xrpl4j/client/XrplClient.java

```diff
@@ -232,6 +238,77 @@ public FeeResult fee() throws JsonRpcClientErrorException {
     return jsonRpcClient.send(request, FeeResult.class);
   }
 
+  /**
+   * Get the ledger index of a tx result response. If not present, throw an exception.
+   *
+   * @return A string containing value of last validated ledger index.
+   */
+  private UnsignedInteger getMostRecentlyValidatedLedgerIndex() throws JsonRpcClientErrorException {
+    JsonRpcRequest request = JsonRpcRequest.builder()
+      .method(XrplMethods.LEDGER)
+      .addParams(LedgerRequestParams.builder().ledgerSpecifier(LedgerSpecifier.VALIDATED).build())
+      .build();
+    return jsonRpcClient.send(request, LedgerResult.class).ledgerIndexSafe().unsignedIntegerValue();
+  }
+
+  /**
+   * Check if the transaction is final on the ledger or not.
+   *
+   * @param transactionHash Hash of the submitted transaction to check the status for.
+   * @return {@code true} if the {@link Transaction} is final/validated else {@code false}.
+   * @throws JsonRpcClientErrorException if {@code jsonRpcClient} throws an error.
+   * @throws InterruptedException if {@link Thread} is interrupted.
+   */
+  public boolean isFinal(Hash256 transactionHash) throws JsonRpcClientErrorException, InterruptedException {
+    JsonRpcRequest request = JsonRpcRequest.builder()
+      .method(XrplMethods.TX)
+      .addParams(TransactionRequestParams.of(transactionHash))
+      .build();
+
+    TransactionResult response = jsonRpcClient.send(request, TransactionResult.class);
+    UnsignedInteger lastLedgerSequence = response.transaction().lastLedgerSequence()
+      .orElseThrow(() -> new IllegalStateException("Ledger result did not contain LastLedgerSequence"));
+
+    // condition 1: validated = true
+    if (response.validated()) {
+      if (response.status().get().equals("success")) {
+        return true;
+      } else {
+        return false;
+      }
+    } else if (FluentCompareTo.is(getMostRecentlyValidatedLedgerIndex()).lessThan(lastLedgerSequence)) {
+      // condition 2: last ledger seq hasn't passed yet
+      Thread.sleep(4000);
+      return isFinal(transactionHash);
```

Comment:
I dont think we want to loop/recurse here. We should just return `false` and let the caller decide how they want to loop.